### PR TITLE
Support Istanbul fork

### PR DESCRIPTION
**Pull request by norhh**
- Implemented 2 opcodes
- Added a precompile
- some gas changes
- Removed python3.5 support as blake2 bindings doesn't support it.(Our own Implementation of blake2 hash would be tedious, slow and would take 200 - 300 lines)

### CODE REVIEW

File: mythril/laser/ethereum/state/environment.py [hunk 1]
```diff
@@ -41,7 +41,10 @@ def __init__(
         self.active_function_name = ""
 
         self.address = active_account.address
+
+        # TODO: Block number should increase with transactions
         self.block_number = symbol_factory.BitVecSym("block_number", 256)
+        self.chainid = symbol_factory.BitVecSym("chain_id", 256)
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/environment.py]: Maybe we should use a default concrete value here? (cli option maybe)
Different chain_id's could in theory have different opcode implementations right?

File: mythril/laser/ethereum/state/environment.py [hunk 2]
```diff
@@ -41,7 +41,10 @@ def __init__(
         self.active_function_name = ""
 
         self.address = active_account.address
+
+        # TODO: Block number should increase with transactions
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/environment.py]: This is not necessarily true. A series of transactions can occur within one block. 
Although there would be a constraint that (new blocknumber >= previous blocknumber)

File: requirements.txt [hunk 3]
```diff
@@ -17,6 +17,7 @@ mock
 persistent>=4.2.0
 plyvel
 py-flags
+py-evm
```
Comment (JoranHonig) [inline requirements.txt]: should we consider version locking?

File: setup.py [hunk 4]
```diff
@@ -19,7 +19,7 @@
 URL = "https://github.com/ConsenSys/mythril"
 AUTHOR = "ConsenSys Dilligence"
 AUTHOR_MAIL = None
-REQUIRES_PYTHON = ">=3.5.0"
+REQUIRES_PYTHON = ">=3.6.0"
```
Comment (JoranHonig) [inline setup.py]: Is this installed by default in Ubuntu now? / Do we have info on the environments of the people that use mythril?

File: mythril/laser/ethereum/state/environment.py (hunk 1 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/environment.py]: (note to self: tox.ini is also changed) 😄

File: mythril/laser/ethereum/natives.py [hunk 5]
```diff
@@ -192,6 +196,19 @@ def ec_pair(data: List[int]) -> List[int]:
...
 
+def blake2b_fcompress(data: List[int]) -> List[int]:
+    """
+    blake2b hashing
+    :param data:
+    :return:
+    """
+    try:
+        parameters = extract_blake2b_parameters(bytes(data))
+    except ValidationError:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/natives.py]: should we log a message here?

Comment (JoranHonig) [reply mythril/laser/ethereum/natives.py]: maybe debug or something?

File: mythril/laser/ethereum/natives.py [hunk 6]
```diff
@@ -192,6 +196,19 @@ def ec_pair(data: List[int]) -> List[int]:
     return [0] * 31 + [1 if result else 0]
 
 
+def blake2b_fcompress(data: List[int]) -> List[int]:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/natives.py]: are there ethereum consensus tests that we could add to check if the implementation is correct?

File: mythril/laser/ethereum/instruction_data.py [hunk 7]
```diff
@@ -210,7 +212,9 @@ def calculate_native_gas(size: int, contract: str):
     elif contract == "identity":
         gas_value = opcodes.GIDENTITYBASE + word_num * opcodes.GIDENTITYWORD
     else:
-        raise ValueError("Unknown contract type {}".format(contract))
+        # TODO: Add gas for other precompiles, computation should be shifted to natives.py
+        #  as some data isn't available here
+        pass
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instruction_data.py]: Now gas_value will contain a None value. Could this break callee's that expect an integer?

File: mythril/laser/ethereum/natives.py (hunk 6 above)
Comment (norhh) [reply mythril/laser/ethereum/natives.py]: There are state tests, but we need to do something with logs to add them. As they check against logs.

Comment (norhh) [reply mythril/laser/ethereum/natives.py]: Also, I have added the fork tests in the tests. So the implementation is correct

File: mythril/laser/ethereum/instruction_data.py (hunk 7 above)
Comment (norhh) [reply mythril/laser/ethereum/instruction_data.py]: Yep, But this was actually fixed in the develop, should merge with it.

File: setup.py (hunk 4 above)
Comment (norhh) [reply setup.py]: People who log issues mainly use docker. So it should be fine. The latest Ubuntu has 3.7.5, only the ancient ones like Ubuntu 16(which is dead) has python 3.5 installed by default.

File: mythril/laser/ethereum/state/environment.py (hunk 1 above)
Comment (norhh) [reply mythril/laser/ethereum/state/environment.py]: Yes, but we can leave it at the moment, as users won't get anything out of it now.

File: mythril/laser/ethereum/natives.py (hunk 6 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/natives.py]: Right I was more wondering about the integration part of it all. To check if the delegation code passes arguments correctly to the function.
The tests that ckeck blake2b directly are nice though 👍

Comment (norhh) [reply mythril/laser/ethereum/natives.py]: I think it tests the delegation trail

**Pull request merged**